drivers/usb/host/dwc_otg/dwc_otg_hcd.c

```diff
@@ -1524,7 +1524,7 @@ int fiq_fsm_setup_periodic_dma(dwc_otg_hcd_t *hcd, struct fiq_channel_state *st,
 		 * state depending on transaction progress.
 		 */
 		blob = (struct fiq_dma_blob *) hcd->fiq_state->dma_base;
-		st->hcdma_copy.d32 = (uint32_t) &blob->channel[hc->hc_num].index[0].buf[0];
+		st->hcdma_copy.d32 = (dma_addr_t) &blob->channel[hc->hc_num].index[0].buf[0];
```

Comment:
This just looks wrong. It at least needs a comment.

Reply:
Looking at it, I can't see how the existing code is correct. ```&blob->....buf[0]``` is going to be a kernel virtual address.

```st->hcdma_copy``` is a ```union hcdma_data```, description
```
/**
 * This union represents the bit fields in the Host DMA Address
 * Register used in Descriptor DMA mode.
 */
..
		/** Start Address of Descriptor List */
		unsigned dma_addr:21;
```

Other places where it is assigned it is given a dma_addr_t
Line 1675:
```
	st->hcdma_copy.d32 = qtd->urb->dma;
```
and ```qtd->urb->dma``` is a ```dwc_dma_t``` (```typedef dma_addr_t dwc_dma_t```)
Line 1854
```
st->hcdma_copy.d32 = hc->align_buff;
```
```hc->align_buff``` is a ```dwc_dma_t```
Except line 1856
```
				st->hcdma_copy.d32 = ((unsigned long) hc->xfer_buff & 0xFFFFFFFF);
```
where ```uint8_t *xfer_buff;```

So ```st->hcdma_copy.d32``` is being treated as a dma_addr_t in most cases, but this one and a couple of others appear to be getting a kernel virtual address. On 64bits that is going to be too big for the 32bit register, so is that ever going to work at the hardware level?

@P33M could you advise?

Reply:
We're casting a 32-bit DMA address (hcd->fiq_state->dma_base) into a "struct fiq_dma_blob *" to do pointer arithmetic. The "fiq_dma_blob" munging is because we're stitching together up to 6 188-byte separate DMA transfers, so instead of porting the DMA pool API into the FIQ we just use array addressing on the DMA base address within a 9kByte contiguous buffer. The buffer is allocated via dma_alloc_coherent() so dma_sync_* calls aren't required.

hc->align_buf is required because of the crippled DMA engine inside the OTG core - it can only DMA to addresses that start at a 4-byte boundary so if we get a URB with a urb->dma address with bottom 2 bits nonzero, a separate buffer is allocated for the core to use.

The DMA address registers are 32 bits wide but the hardware can only address 1G of SDRAM total. If dma_addr_t ends up being 64 bits then we need to guarantee the top 66 bits are 0, or we're going to have a bad time.

Reply:
Thanks. In which case I believe the cast to dma_addr_t is correct, and it wants a comment to explain what the heck is going on.

I think line 1856 ```st->hcdma_copy.d32 = ((unsigned long) hc->xfer_buff & 0xFFFFFFFF);``` should also be a cast to dma_addr_t.
Then again there appears to be some nasty stuff around line 1243 where ```hc->xfer_buff``` being defined as a ```void*``` but gets assigned from ```urb->dma``` (a ```dma_addr_t```), or ```urb->buf``` (a ```void*```) dependent on ```hcd->core_if->dma_enable```. That looks like a module parameter that should have a large warning notice on it.
I think at this point I want to stop looking in case I find further nastiness! It compiles cleanly on both architectures and works in the default configuration, so that's good enough.

dma_addr_t is 64 bits on arm64, but the top 66 bits will be 0 (or is it a 0xc0000000 cache alias?) on BCM283[5|6|7], so all is good.